modules/storages/app/services/storages/file_link_sync_service.rb

```diff
@@ -0,0 +1,218 @@
+      #   failure: result= :error or :not_authorized
+      parse_files_info_response(response)
+    end
+    # Pass errors from Nextcloud into service result
+    @service_result.merge!(nextcloud_request_result)
+
+    if nextcloud_request_result.failure?
+      set_error_for_file_links(storage_file_links)
+      return
+    end
+
+    set_file_link_permissions(storage_file_links, nextcloud_request_result.result)
+  end
+
+  def set_error_for_file_links(storage_file_links)
+    @service_result.result += storage_file_links.each { |file_link| file_link.origin_permission = :error }
+    @service_result.success = false
+  end
+
+  # Check the Nextcloud status of a list of files:
+  # The endpoint returns "canonical" infos per file. Canonical means that the attributes are the same
+  # as the owner sees them and not the current user. If it was the current user the file name could be
+  # different for that user as a shared file can be renamed.
+  # @param file_ids An array of Nextcloud IDs of files to check
+  # @return A HTTP response or an Exception object
+  #
+  # curl -H "Accept: application/json" -H "Content-Type:application/json" -H "OCS-APIRequest: true"
+  # 		-u USER:PASSWD http://my.nc.org/ocs/v1.php/apps/integration_openproject/filesinfo
+  # 		-X POST -d '{"fileIds":[FILE_ID_1,FILE_ID_2,...]}'
+  def request_files_info(file_ids)
+    host = @oauth_client_token.oauth_client.integration.host
+    uri = URI.parse(File.join(host, "/ocs/v1.php/apps/integration_openproject/filesinfo"))
+    request = build_files_info_request(uri, file_ids)
+
+    begin
+      Net::HTTP.start(uri.hostname, uri.port, request_file_info_options) do |http|
+        http.request(request)
+      end
+    rescue StandardError => e
+      e
+    end
+  end
+
+  def build_files_info_request(uri, file_ids)
+    request = Net::HTTP::Post.new(uri)
+    request.body = { fileIds: file_ids }.to_json
+    {
+      'Content-Type': 'application/json',
+      'OCS-APIRequest': 'true',
+      Accept: 'application/json',
+      Authorization: "Bearer #{@oauth_client_token.access_token}"
+    }.each { |header, value| request[header] = value }
+
+    request
+  end
+
+  # HTTP Request options: Keep the request short for the sake of the front-end
+  def request_file_info_options
+    {
+      max_retries: 0,
+      open_timeout: 5, # seconds
+      read_timeout: 3 # seconds
+    }
+  end
+
+  # Takes a response from querying Nextcloud file IDS (an Exception or a HTTP::Response),
+  # parses the returned JSON and
+  # @returns ServiceResult containing data in result and success=true,
+  #   or success=false with result=:error or result=:not_authorized.
+  def parse_files_info_response(response)
+    return ServiceResult.failure(result: :error) if files_info_response_error?(response)
+    return ServiceResult.failure(result: :not_authorized) if response.code == "401" # Nextcloud response if token is not valid
```

Comment:
what about a 403? should be a feasible return value, too, for `:not_authorized`.

Reply:
- 403 is used by Nextcloud (on the per-file level) to indicate missing permissions. This is used for example when a file belongs to another user and is not shared with the current user. That's why there is a 403 cought in set_file_link_permissions.
- 401 is used in Nextcloud OAuth2 to indicate an invalid (expired) token.
I've looked in the PHP source code of Nextcloud to confirm this.

So, 403 could be a feasible return for :not_authorized by Nextclould. However, the current integration_openproject doesn't use it. Other file servers (future implementations) may react differently...

Reply:
hmm, okay. But we could still catch it here, I see no harm in it.

return ServiceResult.failure(result: :not_authorized) if ["401", "403"].include?(response.code)